### ISSUE TITLE
[bug] [enhancement] Excessive piece redownloading when skipping; Fix:Keep last 30s in RAM buffer?

### ISSUE THREAD

**Issue by 21starcade** [elgatito/plugin.video.elementum#837]
First off, the plugin works **amazing** for _continuous playback_ even of 4k videos with just 250-300MB of RAM! Great work, no issues there.

##System
Arm32, Android 7, api 25, 300MB max free RAM, RAM-only storage, Any Kodi version, any video cache setting

##Steps to reproduce/Current behavior
Skipping back while playing videos, **even** just 10 seconds back, causes Kodi to **freeze** playback while Elementum needlessly reconnects peers to **redownload** _the most recently played chunks_ again and again. This is very problematic when you miss a line of dialogue or watch instructional videos where you have to go back and wait repetitively.

-Eg, due to (peering overhead)/(piece availability), skipping back 10s **can cause a 10s freeze** for an 8-mbit video and a 20-25s freeze for a 20-Mbit video despite maximum piece availability, top speeds and having enough RAM to buffer 300/120s of video!

## Expected Behavior
Always protect the _last 30s(or user-set value) played_ in Elementum buffer (~30-75MB reserve) to allow responsive back skipping.

## Possible Solution/Enhancement Pseudocode using a Forward+Recent in-RAM buffer

0. Provide a user option to set an amount of recently played seconds (default:31s) to always be kept in an Elementum "Recent Reserve Buffer", in RAM.
1. Map seconds to torrent pieces and define the Recent Reserve buffer size/(number of pieces) for common bitrates and set the Elementum cache to fit both buffers.
2. Implement a more Pragmatic Download Strategy:
2.0. Prioritize next 30-60s of playback pieces into an Elementum "Forward Buffer" (same as current intent).
2.1. If lots of RAM is free, increase "Recent Reserve Buffer" from 30s to 60s and fill it.
2.2. If free RAM still, keep increasing the "Forward Buffer" with next pieces till RAM is full.
2.3. Move all buffers start times/pieces according to user playback time.

## Context, goals
1. Creating a "Recent" as well as a "Forward" buffer can potentially reserve 1 minute of recent video and up to 3 minutes of upcoming 8-mbit video into a highly responsive, seekable 4-minute buffer on devices with as little as 256MB of free RAM! This should be plenty of cache for all kinds of nearby seeking, except big jumps.

2. Missing a scene or a quick rewind to take notes in-play will be joyfully immediate, as if the entire file is stored.

3. Finding an exact position in a file will be much faster when you can go not just forward but also **backward**.

4. Piece redownloading/peer rediscovery will be minimized, swarms and networks will be healthier, clients will distribute _new pieces_ faster and more frequently.

##Final thoughts and quick fixes
I am not familiar with the interaction between your libtorrent download strategy, the Elementum pieces cache and the Kodi video buffer, BUT an easy fix could be as simple as "do not drop the first buffered pieces" in Elementum's cache. Alternatively, the libtorrent config might be tweaked. Certainly, the Kodi video cache is NOT related.

Hope this helps. :))

**Comment by 21starcade** [comment]
> @21starcade Bitrate is not constant per whole video stream. Elementum is operating in bytes and that is the only possible measurement, not seconds.

Thanks for the clarification, elgatito. As far as I understand, pieces are downloaded and sent to Kodi in 3 stages:

1. libtorrent -> 2. Elementum RAM cache -> 3. Kodi RAM video cache

Currently, once a piece is sent from 2. to 3., it gets immediately dropped from 2. (ie pieces are _**cut**+pasted_ and deleted >playback), which always causes redownloading on backSkip.

Let's say we have an 8-Mbit video, a torrent piece is 16MB on average, Elementum cache is 256MB RAM. Instead of using times/seconds, we can create 32MB of _Recent_ cache by simply _not dropping_ the last 32MB that Elementum passed to Kodi's video cache. This can be expressed as a fraction of user-selected cache size. In this exampple, recent/whole cache = 1/8 (12.5%).

As a function, I imagine something like:
`Keep last 32MB (=2 pieces), sent from Elementum to Kodi video cache, into 1/8 of Elementum's cache.`

Essentially, Elementum would be never deleting the most recently played 32MB of frames from 2. to 3., this way it will creating a 32MB _Recent_ buffer on-the-fly, preventing unnecessary redownloading for the last 30s (for an 8-mbit) video. Then, the usual download strategy (download pieces ahead) can be continued until the other 7/8 of Elementum's cache are filled. Users can tweak the cache size and fraction to match their type of systems/videos.

Can such an option be easily enabled with a simple "don't drop the last 32MB _(or user-set value)_ from ELem cache" code change?

**Comment by elgatito** [comment]
@21starcade 
> Currently, once bytes are sent from 2. to 3., they get immediately dropped from 2. (ie bytes are cut+pasted and deleted >playback), which always causes redownloading on backSkip.

No, piece is removed from memory only when we need to save new piece. Then we remove the oldest piece.
It is usual that Kodi has several active readers of the same file (for example one for video, one for audio stream), so you have few readers and you have to split all the limits/sizes per each reader.

**Comment by 21starcade** [comment]
Did some testing, and interestingly, freezes are more likely _**after the file has been played for a minute**_ , as if there's a problem between the Kodi cache and Elem cache.

When you're seeking quickly with _minimal playback (eg 5s)_ freezes are shorter or less frequent. Is there some sort of waiting period between Kodi and Elem cache when you seek or when a buffer is full?

**Comment by elgatito** [comment]
> When you're seeking quickly with minimal playback (eg 5s) freezes are shorter or less frequent. Is there some sort of waiting period between Kodi and Elem cache when you seek or when a buffer is full?

No, there is no time lag between Kodi and Elementum. 
Kodi is controlling **what** to request from Elementum. Elementum will return requested data if it is already downloaded, or request missing pieces if needed and then Kodi will need to wait once that piece is downloaded.

Elementum cannot know which piece will be requested by Kodi  and when, but Elementum is trying to sequentally download pieces. 

Once you seek player - you will land somewhere and Kodi will request that piece from Elementum and Elementum will try to resort pieces we want to download and invalidate ones that are not required anymore. 

If you want to get back - then your position changes and we will need to resort pieces again. 

Every time you do seeks - we need to download new pieces. With heavy content that takes a while, of course. 

The only good solution to avoid lags while seeking - use file storage, that we are prioritizing pieces but not removing them.
